File: backend/src/main/java/com/softwear/webapp5/model/Transaction.java
package com.softwear.webapp5.model;

import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;

import org.hibernate.annotations.LazyCollection;
import org.hibernate.annotations.LazyCollectionOption;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

	@Column(nullable = false)
    private String type;

	@ManyToOne(optional = false,cascade=CascadeType.REMOVE,  fetch=FetchType.EAGER)
	@OnDelete(action = OnDeleteAction.CASCADE)
    private ShopUser user;

	@ManyToOne (fetch=FetchType.EAGER)
	@OnDelete (action = OnDeleteAction.NO_ACTION)
    private Coupon usedCoupon;

	@Column(nullable = false)
    private String date;

	@Column(nullable = false)
	private Double totalPrice;

	@ManyToMany(cascade=CascadeType.REMOVE, fetch=FetchType.EAGER)
	@OnDelete(action = OnDeleteAction.CASCADE)
	@LazyCollection(LazyCollectionOption.FALSE)
	@Column(nullable = false)
    private List<Product> products;

	public Transaction(String type, String date) {
		this.type = type;
		this.date = date;
	}

    public Transaction(String type, ShopUser user, Coupon usedCoupon, String date, List<Product> products) {
        this.type = type;
        this.user = user;
        this.usedCoupon = usedCoupon;
        this.date = date;
        this.products = products;
		totalPrice = calculateTotalProductPrice();
    }

    public Transaction() {}

	public Transaction(Transaction tr) {
		this.type = tr.type;
        this.user = tr.user;
        this.usedCoupon = tr.usedCoupon;
        this.date = tr.date;
        this.products = tr.products;
		totalPrice = calculateTotalProductPrice();
    }

    public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public ShopUser getUser() {
		return user;
	}

	public void setUser(ShopUser user) {
		this.user = user;
	}

	public Coupon getUsedCoupon() {
		return usedCoupon;
	}

	public void setUsedCoupon(Coupon usedCoupon) {
		this.usedCoupon = usedCoupon;
	}

	public String getDate() {
		return date;
	}

	public void setDate(String date) {
		this.date = date;
	}

	public List<Product> getProducts() {
		return products;
	}

	public void setProducts(List<Product> products) {
		this.products = products;
		totalPrice = calculateTotalProductPrice();
	}

	public Double getTotalPrice() {
		return totalPrice;
	}

	public void setTotalPrice(Double totalPrice) {
		this.totalPrice = totalPrice;
	}

	// Returns sum of product prices, without discount
    public double calculateTotalProductPrice() {
		double sum = 0.0f;
		for(Product product: products) {
			sum += product.getPrice();
		}
		return sum;
	}

}

File: frontend/src/app/shared/classes/transaction.ts
import { ShopUser, Coupon, Product, ProductInTransaction } from '../model';
import {Transaction as TransactionInterface} from '../model/transaction'

export class Transaction implements TransactionInterface{

    id: number;
    type: string;
    user: ShopUser;
    usedCoupon: Coupon;
    date: string;
    totalPrice: number;
    products: Product[];
    productsInTransaction: ProductInTransaction[];

    constructor(transactionInterface: TransactionInterface) {

        this.id = transactionInterface.id;
        this.type = transactionInterface.type;
        this.user = transactionInterface.user;
        this.usedCoupon = transactionInterface.usedCoupon;
        this.date = transactionInterface.date;
        this.totalPrice = transactionInterface.totalPrice;
        this.products = transactionInterface.products;
        this.productsInTransaction = this.initializeProductInTransaction();
    }

    private initializeProductInTransaction(): ProductInTransaction[] {

        const productsInTransaction: ProductInTransaction[] = [];

        this.products.forEach(
            product => {
        
                let productInTransaction = {
                  id: product.id,
                  name: product.name,
                  size: product.size,
                  price: product.price,
                  quantity: 1
                }
        
                let position = this.productIsPresent(productInTransaction) 
                if ( position == -1) {
                    productsInTransaction.push(productInTransaction)
        
                } else {
                    productsInTransaction[position].quantity++
                }
            }
        )

        return productsInTransaction;
    }       
    
    productIsPresent(product: ProductInTransaction): number {

        let isContained = false;
    
        let productIndex = 0;
        if (this.products.length !== productIndex) {
    
            while ((productIndex < this.products.length) && !isContained) {
    
            isContained = this.products[productIndex].id == product.id;
            if (!isContained)
                productIndex++
            }
    
        }
    
        if (!isContained)
            productIndex = -1
    
        return productIndex
    }
}

    
